### PR TITLE
refactor: detach data units from label rows in the project file structure

### PR DESCRIPTION
**Pull request by eloy-encord**
Objectives:
1. Detach data units from label rows in the logic to be able to fetch unannotated data without modifying label row states.
2. Move data units relevant info (not its labels) to the db while not breaking older projects.
3. Refactor ProjectFileStructure to enable calls from it to the db.
4. Update DBConnection to accept different project file structures, not only the default one, in case the user decides to store its data in some other layout (or we change the way the project is structured).

### CODE REVIEW

File: src/encord_active/lib/project/project.py [hunk 1]
```diff
@@ -317,6 +362,19 @@ def split_lr_video(label_row: LabelRow, project_file_structure: ProjectFileStruc
         du = label_row.data_units[data_hash]
         suffix = f".{du['data_type'].split('/')[1]}"
         video_path = (lr_structure.images_dir / du["data_hash"]).with_suffix(suffix)
-        slice_video_into_frames(video_path)
+        sliced_frames, _ = slice_video_into_frames(video_path)
+
+        # 'create_many' is not available for SQLite in prisma
```
Comment (Encord-davids) [inline src/encord_active/lib/project/project.py:367]: [nitpick] 💅

We could probably use `batch` create instead 🙂

File: src/encord_active/lib/file_structure/base.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
+from abc import abstractmethod
+from pathlib import Path
+from typing import Iterator, Optional, Union
+
+from prisma import models
+
+from encord_active.lib.db.base import DataUnitLike
+
+
+class BaseProjectFileStructure:
```
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/file_structure/base.py:10]: Sure, in order to be able to use the database in `ProjectFileStructure` to iterate label rows and data units as we used to do via the filesystem, we need to avoid the cyclic dependency between `ProjectFileStructure` and `PrismaConnection`. That's the main reason why now we add an instance of `BaseProjectFileStructure` to the db connections instead of `project_path` (avoid referencing `ProjectFileStructure` in the db connection code while keeping the type hints and autocompletion). Also, as a collateral is the first step to enable customisable project file structures, not necessarily to enable it for the client but adding flexibility in future file structure changes if we manage to put the type of file structure in a config file.

To sum up:
1. Avoid cyclic dependencies.
2. Adds customisation goddies.

[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/lib/file_structure/base.py:10]: I see, mostly makes sense although I'm personally not a fan of abstraction as a solution to a cyclic dependency, if anything it just hides it away.

If I may suggest another approach, I think composition would be preferable in this case.
Instead of moving the whole class into a base class, we could instead move the logic to iterate label rows and data units into its own class that also depends on `ProjectFileStructure` which seems like a more appropriate separation. wdyt?

File: src/encord_active/lib/project/project.py (hunk 1 above)
Comment (eloy-encord) [reply src/encord_active/lib/project/project.py:367]: fixed, we would want to change the way we have structured the download process as it make more sense to batch in groups of label row + its data units, so we ensure the integrity of the label rows contents in the db.

File: src/encord_active/lib/file_structure/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/file_structure/base.py:10]: IMO more than an abstraction is an interface but we don't have those in python 😄. At some point we started adding logic non-related to the filesystem into `ProjectFileStructure`, so we definitely should separate that logic. Probably we can bear with this one for now, test the robustness of the logic and afterwards migrate it to its own component (while we completely get rid of the old and also not liked behaviour added to `ProjectFileStructure`).

**Review by Encord-davids — APPROVED** [review]
